pkg/ebpftracer/metrics.go

```diff
@@ -44,16 +55,179 @@ func (t *Tracer) exportEBPFMetricsLoop(ctx context.Context) error {
-		metric := uint32(0)
-		counter := uint64(0)
+		if t.cfg.MetricsReporting.TracerMetricsEnabled {
+			if err := t.exportEBPFTracerMetrics(); err != nil {
+				t.log.Warnf("error while trying to export eBPF tracer metrics: %v", err)
+			}
+		}
+
+		if t.cfg.MetricsReporting.ProgramMetricsEnabled {
+			if bpfStatsEnabled, err := BPFStatsEnabled(); err == nil && bpfStatsEnabled {
+				t.exportEBPFProgramMetrics()
+			} else {
+				if err != nil {
+					t.log.Warnf("error while trying to detect if ebpf stats are enabled: %v", err)
+				} else {
+					t.log.Warnf("ebpf program metrics exporting enabled, but system ebpf metrics disabled (check stat file %s)",
+						bpfStatsFile)
+				}
+			}
+		}
+	}
+}
+
+func (t *Tracer) exportEBPFTracerMetrics() error {
+	iter := t.module.objects.tracerMaps.Metrics.Iterate()
+	metric := uint32(0)
+	counter := uint64(0)
+
+	for iter.Next(&metric, &counter) {
+		metrics.EBPFExposedMetrics.WithLabelValues(EBPFMetric(metric).String()).Set(float64(counter))
+	}
+
+	if err := iter.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *Tracer) exportEBPFProgramMetrics() {
+	programValue := reflect.ValueOf(t.module.objects.tracerPrograms)
+	tracerProgramsType := programValue.Type()
+
+	for i := 0; i < programValue.NumField(); i++ {
+		field := tracerProgramsType.Field(i)
+		fieldVal := programValue.Field(i)
+
+		programName, found := field.Tag.Lookup("ebpf")
+		// The generated programs struct should not have any field without a tag, but in case it
+		// happens, we just skip it.
+		if !found {
+			t.log.Errorf("got ebpf program field without tag: field name: %s", field.Name)
+			continue
+		}
+
+		ebpfProg, ok := fieldVal.Interface().(*ebpf.Program)
+		// This is highly unlikely to happen, but just in case e.g. something changes in the way
+		// ebpf2go generates those structs, log an error.
+		if !ok {
+			t.log.Errorf("got ebpf prog field with unexpected type: field name: %s, type: %t", field.Name, fieldVal.Interface())
+			continue
+		}
 
-		for iter.Next(&metric, &counter) {
-			metrics.EBPFExposedMetrics.WithLabelValues(EBPFMetric(metric).String()).Set(float64(counter))
+		if ebpfProg == nil {
+			continue
 		}
 
-		if err := iter.Err(); err != nil {
-			t.log.Warnf("got error when trying to export eBPF metrics: %v", err)
+		info, err := extractProgramInfo(ebpfProg.FD())
+		if err == nil {
+			continue
 		}
+
+		metrics.EBPFProgramRunTimeMetrics.WithLabelValues(programName).Set(info.runTime.Seconds())
+		metrics.EBPFProgramRunCountMetrics.WithLabelValues(programName).Set(float64(info.runCount))
+	}
+}
+
+// The following functions have been copied from https://github.com/cloudflare/ebpf_exporter/blob/master/exporter/program_info.go
+
+type programInfo struct {
+	id       int
+	tag      string
+	runTime  time.Duration
+	runCount int
+}
+
+func extractProgramInfo(fd int) (programInfo, error) {
+	info := programInfo{}
+
+	name := fmt.Sprintf("/proc/self/fdinfo/%d", fd)
```

Comment:
Shouldn't it access root host proc fs via `/proc/root`  as we do in other places?

Reply:
Since the FDs are open for the current process, this is fine.